### ISSUE TITLE
IREE's llvm-cpu backend miscompiles stablehlo.transpose operations with i4 operands and odd-dimensioned columns or rows.

### ISSUE THREAD

**Issue by phoenix-meadowlark** [iree-org/iree#16219]
### What happened?

IREE miscompiles `stablehlo.transpose` operations with operands that
1. have or had `i4` types and 
2. have an odd row or column dimension.

The resulting array is incorrect in every other column, and contains values incorrectly indexed one element too early relative to the flattened input. For example if the result was supposed to be:

```
[[0 3
  1 4
  2 5]]
```

then the result would be:

```
[[0 2
  1 3
  2 4]]
```

### Steps to reproduce your issue

The `constant` below is about as small as I could make it without IREE throwing a compilation error.

```mlir
// /tmp/transpose.mlir
module @jit_transpose attributes {mhlo.num_partitions = 1 : i32, mhlo.num_replicas = 1 : i32} {
  func.func public @main() -> (tensor<13x42xi8> {jax.result_info = "[0]", mhlo.layout_mode = "default"}) {
    %0 = stablehlo.constant dense<"0x08090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F000102030405060708090A0B0C0D0E0F00010203040506070809"> : tensor<42x13xi4>
    %1 = stablehlo.convert %0 : (tensor<42x13xi4>) -> tensor<42x13xi8>
    %2 = stablehlo.transpose %1, dims = [1, 0] : (tensor<42x13xi8>) -> tensor<13x42xi8>
    return %2 : tensor<13x42xi8>
  }
}
// converting to i8 after the transpose produces the same miscompilation.
```

```shell
iree-compile \
  --iree-hal-target-backends=llvm-cpu \
  --iree-input-type=stablehlo \
  /tmp/transpose.mlir \
  -o /tmp/transpose.vmfb
```

The following compilation warning is thrown:

```mlir
/tmp/transpose.mlir:9:10: warning: skipping consteval initializer: unsupported type for current jit configuration: 'tensor<42x13xi4>'
    %1 = stablehlo.transpose %0, dims = [1, 0] : (tensor<42x13xi4>) -> tensor<13x42xi4>
         ^
/tmp/transpose.mlir:2:3: note: called from
  func.func public @main() -> (tensor<13x42xi8> {jax.result_info = "[0]", mhlo.layout_mode = "default"}) {
  ^
```

For smaller `stablehlo.constants`, the following error is thrown:

```mlir
<unknown>:0: error: failed to legalize operation 'memref.get_global' that was explicitly marked illegal
<unknown>:0: note: see current operation: %7 = "memref.get_global"() <{name = @__constant_42x11xi4}> : () -> memref<42x11xi4>
/tmp/lookup_emb/simplified.mlir:5:10: error: 'builtin.module' op failed to emulate bit width
    %2 = stablehlo.transpose %1, dims = [1, 0] : (tensor<42x11xi8>) -> tensor<11x42xi8>
         ^
```

```shell
iree-run-module \
  --device=local-task \
  --function=main \
  --module=/tmp/transpose.vmfb
```

```
xla_result:
[[-8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3]
 [-7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2]
 [-6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1]
 [-5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0]
 [-4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1]
 [-3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2]
 [-2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3]
 [-1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4]
 [ 0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5]
 [ 1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6]
 [ 2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7]
 [ 3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8]
 [ 4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7  6  3  0 -3 -6  7  4  1 -2 -5 -8  5  2 -1 -4 -7]]

iree_result:
[[-8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4]
 [-7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3]
 [-6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2]
 [-5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1]
 [-4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0]
 [-3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1]
 [-2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2]
 [-1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3]
 [ 0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4]
 [ 1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5]
 [ 2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6]
 [ 3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7  5  1 -1 -5 -7  5  3 -1 -3 -7  7  3  1 -3 -5  7]
 [ 4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8  6  2  0 -4 -6  6  4  0 -2 -6 -8  4  2 -2 -4 -8]]

xla_result - iree_result:
[[  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]
 [  0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0 -15]
 [  0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1   0   1]]
```

### What component(s) does this issue relate to?

Compiler, Runtime

### Version information

eef24a93ed6235a12828aef6ab94f3ae629f1aab

**Comment by benvanik** [comment]
is this from an actual program or a test suite?

odd dims in sub-byte element typed tensors are likely going to be reaaaally hard to support to the point of it likely being something we need to guard on and fail compilation (or emit runtime asserts on the host side for dynamic cases). there's an assumption (met by all shipping models) that there's at least byte alignment on each dimension of sub-byte element typed tensors. static shapes may be possible to emit a bunch of extra code to handle but dynamic shapes will really bloat things trying to support all possible paths where there may be byte-unaligned elements.

**Comment by phoenix-meadowlark** [comment]
This is from the main model that my team is working on, unfortunately. I obfuscated the original size but the weight has an odd and an even dimension.

**Comment by benvanik** [comment]
for performance reasons you'll likely want to change the model :)
(mostly critical if on an innermost dim, but can't always assume an outermost won't end up as an innermost as a transpose is added)

**Comment by benvanik** [comment]
(if we can fix this that'd be great - I'll let someone else chime in - just pointing out that there's likely to be a lot of pitfalls around doing this and it's unlikely to get a lot of attention because it's an anti-pattern with packed types like this)

**Comment by phoenix-meadowlark** [comment]
> for performance reasons you'll likely want to change the model :) (mostly critical if on an innermost dim, but can't always assume an outermost won't end up as an innermost as a transpose is added)

I'll look into it in the meantime! if I separately pre-process the checkpoint to pad the weight, integrating it with the rest of the infrastructure should hopefully be straightforward.

**Comment by cc10512** [comment]
> for performance reasons you'll likely want to change the model :) (mostly critical if on an innermost dim, but can't always assume an outermost won't end up as an innermost as a transpose is added)

True, but if the user picks that dimension, it would be very helpful for the compiler to generate correct code and let the user worry about performance. Or even better, print a message warning them that the performance can be improved by using dimensions that pack nicely in native machine types.

**Comment by benvanik** [comment]
Yep - if we can make it work that'd be great - but as soon as you have a single dynamic dimension generating (effectively) entire copies of all dispatches to handle such cases is not really desirable. I suspect things will just explode in those cases today, and we'd probably need to add a global flag ala clang's `-mno-unaligned-access` in order to have 99% of programs perform well/compile faster (and given our domain/what others are doing we'd default that flag to on).

**Comment by hanhanW** [comment]
> For smaller stablehlo.constants, the following error is thrown:
```mlir
<unknown>:0: error: failed to legalize operation 'memref.get_global' that was explicitly marked illegal
<unknown>:0: note: see current operation: %7 = "memref.get_global"() <{name = @__constant_42x11xi4}> : () -> memref<42x11xi4>
/tmp/lookup_emb/simplified.mlir:5:10: error: 'builtin.module' op failed to emulate bit width
    %2 = stablehlo.transpose %1, dims = [1, 0] : (tensor<42x11xi8>) -> tensor<11x42xi8>
```

This is a known issue. I did not fix it because it is very rare that we have small i4 constants. If you want to make it work, you need to add a pattern to [EmulateNarrowType](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/EmulateNarrowType.cpp). You also need to create packed constant at MLIR level. Because MLIR uses an i8 storage for each i4 element.

For actual issue, you'll need someone to look into IRs and generated code to figure out what's happening.

Pro tip to debug it with small constants: you can add `--iree-flow-inline-constants-max-byte-length=0` to iree-compile. Then all the small constants won't be pulled into dispatch.

**Comment by phoenix-meadowlark** [comment]
I was able to circumvent this issue by modifying the dimension in the model, which provided a modest speedup :)

My vote is that we handle this by making cases like this raise compiler errors. Hunting down miscompilations via numerical differences is pretty unpleasant.
